opencog/nlp/sureal/SuRealModule.cc

```diff
@@ -129,13 +124,106 @@ HandleSeq SuRealModule::do_sureal_match(Handle h)
         sVars.insert(n);
     }
 
-    logger().debug("[SuReal] starting pattern matcher");
+    // separate the disconnected clauses (this will happen often with SuReal)
+    std::set<HandleSeq> connectedClauses;
+    get_connected_components(sVars, qClauses, connectedClauses);
+
+    logger().debug("[SuReal] Found %d disconnected components", connectedClauses.size());
+
+    std::map<Handle, std::vector<std::map<Handle, Handle> > > collector;
+
+    // call the pattern matcher on each set of disconnected commponents
+    for (auto& c : connectedClauses)
+    {
+        logger().debug("[SuReal] starting pattern matcher");
+
+        // copy the clause for passing const stuff to non-const argument list
+        HandleSeq qClause(c);
+
+        SuRealPMCB pmcb(pAS, sVars);
+        m_pme.match(&pmcb, sVars, qClause, qNegs);
+
+        // no pattern matcher result
+        if (pmcb.m_results.empty())
+            return HandleSeq();
+
+        // first disconnected component & result? add it all
+        if (collector.empty())
+        {
+            collector = pmcb.m_results;
+            continue;
+        }
+
+        // if we are checking subsequent disconnected components, we want to
+        // make sure the new results can be merged with results from previously
+        // checked components; ie. only keep those with common
+        // InterpretationNode & no overlapping mappings
+        for (auto it = collector.begin(); it != collector.end(); )
+        {
+            // no common Interpretation, erase the old results as it can no
+            // longer be satisfied
+            if (pmcb.m_results.count(it->first) == 0)
+            {
+                logger().debug("[SuReal] Discarding a result for %s", it->first->toShortString().c_str());
+
+                it = collector.erase(it);
+                continue;
+            }
+
+            auto& existingMaps = it->second;                // a vector of previous mappings
+            auto& appendMaps = pmcb.m_results[it->first];   // a vector of unmerged mappings
+            std::vector<std::map<Handle, Handle> > newMaps;
+
+            // check all combinations of all the old mappings to the new
+            for (auto& em : existingMaps)
+            {
+                for (auto& am: appendMaps)
+                {
```

Comment:
Oh never mind. You write good code, and I talk before I think. Yes, if you accumulate everything done so far into "existing", and then add on to each of those, it will work.  So yes, this looks right after all. Sorry...